### ISSUE TITLE
Grades do not match expectations for everything but osu!mania

### ISSUE THREAD

**Issue by Ratchet0203** [ppy/osu#4184]
While trying out all the modes to see the new look of the lazer skin, I noticed that during breaks and on result screens, the wrong letter grade showed. To prove I'm understanding of the scales for each game mode, I'll list them below.
osu!standard/osu!taiko:
SS - 100% accuracy
S - >=90% 300s, 0 miss (<=1% 50s in standard)
A - >=90% 300s with misses/80-90% 300s with no misses
B - 80-90% 300s with misses/70-80% 300s with no misses
C - 60-80% 300s with misses/60-70% 300s with no misses
D - Less than 60% 300s, misses or not.
osu!catch
SS - 100% accuracy
S - 98% accuracy
A - 94% accuracy
B - 90% accuracy
C - 85% accuracy
D - Less than 85% accuracy
osu!mania
SS - 100% accuracy
S - 95% accuracy
A - 90% accuracy
B - 80% accuracy
C - 70% accuracy
D - Less than 70% accuracy
In all modes, HD and/or FL mods will change S and SS to S+ and SS+, also known as silver S and silver SS. However, every single mode follows that of mania's scale. Also, instead of sliders being counted as one hitburst, sliderends and sometimes sliderscorepoints are counted as extra acc. I dislike this, because the total number of hitbursts no longer add up to the number of objects in the map. Spinners also don't net bonus points after clearing, and even with mods, max score appears to still be 1,000,000, as if no mods are active at all.

**Comment by HoLLy-HaCKeR** [comment]
Duplicate of #4002

**Comment by peppy** [comment]
Thanks for pointing this out. I don't believe this is a duplicate as it is focusing on accuracy/rank of all rulesets.

With lazer, we have been going for conformity across rulesets, so it is quite possible that things will *actually remain* how they are and not match stable. This is still an ongoing discussion, but I don't believe we will be making any rash changes to the cutoffs for ranks just yet (and it we do, it would be applied across all rulesets).

**Comment by MatterMr** [comment]
I am curious on how FC plays are considered in letter grades for Lazer. Some of the charm of getting an "S" rank in stable has been that it represents a play that has no misses. The intuitive solution in my mind was to append a "+" to the letter grade to represent a FC, but this seems to be currently implemented as the benefit for submitting an FL/HD play. I think keeping the silver coloring, and changing the meaning of the "+" would be a better solution to representing a FC play.

**Comment by bdach** [comment]
@MatterMr new grades are likely out of scope of this issue.

**Comment by monochrome22** [comment]
For standard, the main issue that people seem to have is not the accuracy ranges, but rather that an FC is not distinguished, so the solution for standard is to figure that out. Does that still make grade changes out of scope for this issue?

**Comment by Zyfarok** [comment]
Going back to osu! stable grades, and simply removing the weird "(<=1% 50s in standard)" condition from S-ranks should be enough.

If something new wants to be tried, Acc could be used instead of 300s ratio (because that part is slightly weird ?), or alternatively Score could be used, but the most important is that S definitely needs to remain a "FC with good accuracy" and the A/B/C/D needs to be closely tied to accuracy. Whether a low acc FC should be treated like non-FCs or be shifted by one grade (as in stable) is debatable.

**Comment by Ratchet0203** [comment]
I think the grades should just be kept the same as stable. There's no need to overcomplicate things, and one of the biggest gripes people have with lazer (myself included) is how different everything is.

**Comment by monochrome22** [comment]
> Going back to osu! stable grades, and simply removing the weird "(<=1% 50s in standard)" condition from S-ranks should be enough.

Keep in mind that going back to stable's system would mean the grade system couldn't be as easily explained as it is now, by displaying the graph at results screen,

I said this in another thread, but I'll also say it here for visibility: personally my only issue with grades is that you can't tell the difference between FC and not FC. I appreciate the more simple grading system, especially because even after years of playing stable, the grade system was a mystery (osu! game mode).

**Comment by Zyfarok** [comment]
S has always been associated with FC so it would break too much assumptions from stable people and confuse them to have S that are not FCs.
Also, the S+ notation has been used for silver S (S with HD/FL) in a lot of skins which would increase the confusion.

Changing the other grades and introducing separate notations for low-acc FCs is probably fine though.

**Comment by developomp** [comment]
I think this is a perfect opportunity to make the S rank only available for FC plays.

**Comment by Ladwick** [comment]
> S has always been associated with FC so it would break too much assumptions from stable people and confuse them to have S that are not FCs

Well, in stable not FCs S still exist thanks to sliderbreaks, I get that having a S from a score with a miss may be kind of weird but gaining ruleset parity is a good point to me.

Still, I have a personal perspective regarding the whole topic that I would like to share here:

SSS -> 100% acc
SS -> 95%+ acc (no misses but still doable with sliderbreaks and no completing all sliderends)
S -> 95%+ acc (w/ misses)

And their respective SSSH, SSH and SH variants.

**Comment by peppy** [comment]
We're not adding a new grade at this point in time (and probably ever). That's all I'll say about the previous comment.

**Comment by stockfishcooker** [comment]
I have a suggestion: give a new symbol for FCs (something like a star). This keeps the "+" sign for FL/HD submissions.
As for the grading system, I think that the stable system should be kept for classic mod only. While it is a part of osu!'s history, I think that having a more intuitive system would really help out newer players a lot.

**Comment by Ladwick** [comment]
To be honest I don't think the classic grading system is soo intuitive compared to the new one because even sometimes even if you don't miss, the chance of getting an S would depend on a % of judgements in the whole map, I really think you can't have a more intuitive grading system than the current one because is so simple as 100%> SS, 95%> S, 90%> A and so on, plus, getting an S in those terms is already harder than in stable classic gameplay thanks to sliderhead judgment.

The [PERFECT] is always there to verify if you FCed the map, but making the perfect more flashy like part of the grade or like a big word or symbol next to the grade would be good yeah

**Comment by bdach** [comment]
I'm not really seeing a clear path to getting this issue out of discussion. The main things for me here are:

- What does the playerbase even want out of grades? (hi @Bubbleman532)
- If we were to make the grading system any more complicated than a percentage threshold, then what do we do with all the designs that rely on the grading system being presentable using a simple pie chart? (hi @arflyte)

**Comment by peppy** [comment]
I'll attempt to summarise everything in this thread so far, and to give an understanding of where things currently are.

Grades are given out by the following rules:

<table>
    <tr>
        <th>grade
        <th>osu! / osu!taiko
        <th>osu!catch
        <th>osu!mania
        <th>lazer (all rulesets)
    <tr>
        <th>SS
        <td>100%
        <td>100%
        <td>100%
        <td>100%
    <tr>
        <th>S
        <td>≥90% (≤0.1% 50s, 0% miss)
        <td>≥98%
        <td>≥95%
        <td>≥95%
    <tr>
        <th>A
        <td>≥80%FC or ≥90%
        <td>≥94%
        <td>≥90%
        <td>≥90%
    <tr>
        <th>B
        <td>≥70%FC or ≥80%
        <td>≥90%
        <td>≥80%
        <td>≥80%
    <tr>
        <th>C
        <td>≥60%
        <td>≥85%
        <td>≥70%
        <td>≥70%
</table>

Some summary based on the above and the discussion so far:

- Lazer matches osu!mania perfectly (no issue)
- osu!catch has adjusted cutoffs because it has a very different average spread of judgements. This definitely requires some kind of consideration, whether that comes from changing how accuracy is handled or from changing cutoffs to match community expectations. That said, we haven't had much feedback on this yet.
- osu! and osu!taiko both use an arcane judgement system that is far too complex, but ignoring these edge cases, it roughly matches below 90% accuracy.
- I don't think anyone is going to be against removing the `xx%FC or xx%` conditionals on B and A grades, so let's assume those are a non-issue to standardise away.
- The largest point of feedback and contention is how full combos should play into the SS/S/A rank range for osu!/osu!taiko. Based on feedback so far I think this definitely needs to be a consideration going forward.

So... I don't think there is a chance of standardising the grades across all rulesets. Here's a proposal to get further discussion started:

<table>
    <tr>
        <th>grade
        <th>osu! / osu!taiko
        <th>osu!catch
        <th>osu!mania
    <tr>
        <th>SS
        <td>100%
        <td>100%
        <td>100%
    <tr>
        <th>S
        <td>≥95%FC
        <td>≥98%
        <td>≥95%
    <tr>
        <th>A
        <td>≥90%
        <td>≥94%
        <td>≥90%
    <tr>
        <th>B
        <td>≥80%
        <td>≥90%
        <td>≥80%
    <tr>
        <th>C
        <td>≥70%
        <td>≥85%
        <td>≥70%
</table>


Note that osu!catch and osu!mania match stable implementations in this proposal. I'd love to standardise osu!catch more based on feedback (ie. can we change S grade to `>95%FC` there as well? or is that too easy?)

**Comment by developomp** [comment]
Even if osu!catch is standardized to be more consistent with other game modes, that still leaves osu!mania different from others as it does not require an FC for S rank.

I think it's best we leave osu!catch as it is since it is such a drastic change for players for a little bit of consistency.

**Comment by Pinossaur** [comment]
I wouldn't be opposed to trying to standardize game modes as much as possible, but I don't think catch grades requirements can't be lowered further, as catch doesn't have a tapping mechanic, and thus accuracy doesn't play as much of a role. I could be wrong, but I believe unless people are specifically doing an effort to not catch "slider fruit" (sorry for not knowing the right terminology) which from what I understand are the only way of dropping accuracy without dropping combo or HP. 

In a nutshell I don't think that scores in catch will drop much below 85% unless played with NF, as dropping accuracy while passing a map is something harder to do.

**Comment by heyngra** [comment]
> I'll attempt to summarise everything in this thread so far, and to give an understanding of where things currently are.
> 
> Grades are given out by the following rules:
> 
> <table>
>     <tr>
>         <th>grade
>         <th>osu! / osu!taiko
>         <th>osu!catch
>         <th>osu!mania
>         <th>lazer (all rulesets)
>     <tr>
>         <th>SS
>         <td>100%
>         <td>100%
>         <td>100%
>         <td>100%
>     <tr>
>         <th>S
>         <td>≥90% (≤1% 50s, 0% miss)
>         <td>≥98%
>         <td>≥95%
>         <td>≥95%
>     <tr>
>         <th>A
>         <td>≥80%FC or ≥90%
>         <td>≥94%
>         <td>≥90%
>         <td>≥90%
>     <tr>
>         <th>B
>         <td>≥70%FC or ≥80%
>         <td>≥90%
>         <td>≥80%
>         <td>≥80%
>     <tr>
>         <th>C
>         <td>≥60%
>         <td>≥85%
>         <td>≥70%
>         <td>≥70%
> </table>
> 
> Some summary based on the above and the discussion so far:
> 
> - Lazer matches osu!mania perfectly (no issue)
> - osu!catch has adjusted cutoffs because it has a very different average spread of judgements. This definitely requires some kind of consideration, whether that comes from changing how accuracy is handled or from changing cutoffs to match community expectations. That said, we haven't had much feedback on this yet.
> - osu! and osu!taiko both use an arcane judgement system that is far too complex, but ignoring these edge cases, it roughly matches below 90% accuracy.
> - I don't think anyone is going to be against removing the `xx%FC or xx%` conditionals on B and A grades, so let's assume those are a non-issue to standardise away.
> - The largest point of feedback and contention is how full combos should play into the SS/S/A rank range for osu!/osu!taiko. Based on feedback so far I think this definitely needs to be a consideration going forward.
> 
> So... I don't think there is a chance of standardising the grades across all rulesets. Here's a proposal to get further discussion started:
> 
> <table>
>     <tr>
>         <th>grade
>         <th>osu! / osu!taiko
>         <th>osu!catch
>         <th>osu!mania
>     <tr>
>         <th>SS
>         <td>100%
>         <td>100%
>         <td>100%
>     <tr>
>         <th>S
>         <td>≥95%FC
>         <td>≥98%
>         <td>≥95%
>     <tr>
>         <th>A
>         <td>≥90%
>         <td>≥94%
>         <td>≥90%
>     <tr>
>         <th>B
>         <td>≥80%
>         <td>≥90%
>         <td>≥80%
>     <tr>
>         <th>C
>         <td>≥70%
>         <td>≥85%
>         <td>≥70%
> </table>
> 
> 
> Note that osu!catch and osu!mania match stable implementations in this proposal. I'd love to standardise osu!catch more based on feedback (ie. can we change S grade to `>95%FC` there as well? or is that too easy?)

I think that the no miss to S grade should be still added.
Why?
When you check the rank, when you see the leaderboards there are people who have lower acc, higher acc, lower combo (slider ends and stuff), FCs. Without looking exactly on the replay, or on the score sheet, you can't really quickly tell if someone did actually FC or not (in the meaning of not missing a combo, counting sliderends into it), or if they missed at some point. 
Also, the S grade should be more rare than A grade, more unique. If we have SS as "do 100% acc, be perfect", then the S rank should also have some condition that makes it unique from other, normal grades as A,B and so on, and that shows you are not THAT perfect player as SS on the map, yet you didnt miss, so you're getting here! imo, the no miss requirement is pretty easy to understand, and is recognizable enough to make the **S** rank actually **S**pecial

**Comment by peppy** [comment]
> I think that the no miss to S grade should be still added.

This is already in the proposal. If you agree with that, you can leave a reaction rather than writing  along reply 😉 .

**Comment by ominoussage** [comment]
> S	≥95%FC

I'm a little bit confused. Does this mean that both criteria have to be met to obtain an S grade? One thing I loved about lazer is that 95% and above is graded as S, regardless if it's an FC or not. It would suck seeing an A graded 99% on top of S graded 96% FCs. 

I think S graded plays should be superior than A graded plays in all aspects and should always be above A graded plays (when comparing plays that use the same mods).

**Comment by Nikku514** [comment]
From a catch players perspective the only part of the >95%FC I'd agree with is tying S to if it's an FC. You would need to be dodging/ignoring accuracy or it being a CS6+ map with a lot of sliders to actually get below 97-98% on most FCs. I will say, 98% S is actually very forgiving for the majority of maps in my opinion, if you need an example of why I say that just look at the average profile accuracy everyone has.
The best change from how I see it is S being >98%FC but right now I honestly really do not think catch should be changed.

The reason I did not mention A/B/C is because they are in pretty good places, only thing to mention is A _could_  be 95% but that's just my brain liking multiples of 5.

**Comment by Okorin** [comment]
I think osu/taiko rulesets following >95% = S would be fine too, would bring forward the idea of accuracy mattering more i think and instant A rank is just frustrating esp in taiko if you're just average at the game

**Comment by rszyma** [comment]
(An average osu!standard player's opinion)

> S ≥95%FC

I pretty sure that most standard players associate an S rank score more with a FC (or rather 0 miss) rather than high accuraccy. Therefore I think S rank fits better the 0 miss scores.

However I propose to decrease required percentage for S from 95% to 90%. So the S would be ≥90%+FC (or 0 miss) and A rank would be also ≥90% but no FC (0 miss). That's because osu!standard puts a very heavy weight on miss count (in terms of pp). Also 90% for S would be closer to how S rank accuraccy requirement for stable.

The proposed by peppy B and C standard score requirements scores lgtm.

**Comment by Ladwick** [comment]
> S ≥95%FC

Regarding stable, I'm pretty sure many people is not going to like this, but I consider it a necessary change if the 50 acc/50 combo or 70 acc/50 combo score change is happening. Of course it will be harder to get an S but it's the ideal chance to let the community know that accuracy will be more important towards the new 'meta' that lazer will bring.

**Comment by KcHecKa** [comment]
in lazer, getting a 100% in mania means you need a perfect 1,000,000 score, which is very different from how it behaves in stable...

**Comment by bdach** [comment]
@KcHecKa #24238

**Comment by plaguemask** [comment]
One thing that I feel shouldn't be left unsaid: The table for stable isn't exactly correct, for a couple reasons. One is sliderbreaks, which [peppy already touched on](https://twitter.com/ppy/status/1704068462639992893).

The other one is that the osu! gamemode's percentage thresholds *aren't* actually based on accuracy. It's based on *the percentage of 300s.* This means, from a pure accuracy perspective, the threshold between A and B can be as low as 90% accuracy (10% misses) and as high as 93.33% (10% 100s, close to 0% misses), or anywhere in between. [This response](https://twitter.com/onlyatroller/status/1704091990474830198) to peppy's acc tweet refers to this obliquely, and matches my experience of playing osu! over multiple years and internalizing that the cutoff between A and B (or S and A w/out misses) is usually around 93%. Additionally, a long while ago, after looking into osu!'s acc system and seeing how unintuitive it is, I made [this infographic](https://cdn.discordapp.com/attachments/97787876338577408/1153900452010807356/osu_grades_by_accuracy.png) explaining its weirdness.

It's a no-brainer to just use accuracy going forward, and the difference it makes isn't very large, but I felt this discrepancy should at least be noted here for the record.

**Comment by peppy** [comment]
The table is correct. Please re-read the conditions stated. You can re-order and try to aggregate the conditions all you want, but this isn't going to help with explaining. The conditions in the table are what we use in code. Knowing that "it's actually 90-93.33438%" isn't going to help, as you seem to be aware of. So I'd ask that you keep these kinds of replies out of this thread just to keep noise to a minimum.

**to all future commenters: if you're here to agree or disagree, please use reactions rather than posting**

**Comment by Nyxa** [comment]
I've been thinking about Lazer grades for a while, ever since I started playing Lazer with the updated grading system. I've always found it interesting to think about how the focus on accuracy for grading changed my approach to playing, but I also wound up missing having some sort of indicator for FCs. 

Additionally, as the game has evolved, the Silver S/SS have become rather redundant, in that while they stand out to look special, people are generally unimpressed with HD scores (and FL is extremely rare to see an FC with beyond low SR maps anyway) and so the different colours of grades become visual noise; they're not really all that special, especially considering there's many players who always have HD on because they find it easier. 

Below is more or less what I wound up coming up with after discussing these ideas with some other players.

| Accuracy  | No FC | FC
| ------------- | ------------- |  ------------- |
| 100%  | N/A | SS (Silver/Gold?) |
| 98%+ | S (Gold) | S+ (Gold) |
| 95%+ | S (Silver) | S+ (Silver) |
| 90%+ | A | A+ |
| 80%+ | B | B+ |
| 70%+ | C | C+ |
| <70% | D | D+ |

In this scenario, the `+` symbol can be replaced with anything - it's symbolic for some kind of indicator that represents an FC, that's easily recognizable but orthogonal to the actual grade itself, which is purely accuracy based. Additionally, an additional tier of accuracy is introduced, so that very high acc scores are put in their own bracket, since there _is_ a pretty major difference between a 98% score and a 95% score.

The only thing I haven't figured out what to do with is the golden/silver variants of the SS. I figure silver should be what you get for a nomod SS, and golden SS would be granted if you SS the map with any mods, or at the very least SR-increasing mods (so HR/DT). That's what would make the most sense to me, but I don't know how others would feel about this.

**Comment by peppy** [comment]
> In this scenario, the + symbol can be replaced with anything - it's symbolic for some kind of indicator that represents an FC

There's a large "Perfect" text which represents a full combo. This will be added back in lazer. It should not be part of the grade.

**Comment by Nyxa** [comment]
> > In this scenario, the + symbol can be replaced with anything - it's symbolic for some kind of indicator that represents an FC
> 
> There's a large "Perfect" text which represents a full combo. This will be added back in lazer. It should not be part of the grade.

Would this be visible on the scores listing or would you need to click a score to see it?

**Comment by Morroni** [comment]
There are so many extremely good plays from top players where they have only 1 miss but 99%+ accuracy. I think such plays worth being S, not A. 

For people, who are trying to say that S rank should represent FC: why there is a S rank on plays with 93~94% accuracy and a couple of sliderbreaks? I think that doesn't make any sense either. 

The system used in osu!mania is correct and suitable for osu!std mode as well. Prove me wrong.

**Comment by agatemosu** [comment]
> There are so many extremely good plays from top players where they have only 1 miss but 99%+ accuracy. I think such plays worth being S, not A. 

With that in mind, I think the table could be modified like this

| Grade |	Condition |
| SS | 	100%	|
| S |	≥95%FC	or ≥99% |
| A	| ≥90% |
| B |	≥80%	|
| C |	≥70% |

**Comment by monochrome22** [comment]
> I guess it's a valid question: If "perfect" is shown on all leaderboards in some way, are people willing to have the accuracy cutoffs match what lazer currently has?

2 issues with this: players wouldn't be rewarded for FCing (any text or color wouldn't be nearly as rewarding as a grade) and if someone said "I have an S rank on this map", you wouldn't know if it's an FC or not. 

Personally I still support using +, but if there is a good visual design for results screen to explain that S rank is ≥95%FC, then that would be the next best solution.

**Comment by peppy** [comment]
Using `+` grades is **out of the question**. This doesn't not mean FC. Other games use it for better accuracy. We are not giving it a new meaning, so please drop that idea.

I'm leaning towards adjusted accuracy cutoffs and better FC visibility right now. Seems like a good compromise.

**Comment by Okorin** [comment]
FWIW nyxa's suggestion only mentioned the `+` as a placeholder for any FC indicator that is easily distinguishable, so at least in that regard thats on the same page, im more interested if repurposing the silver S to be an inferior version of the golden one is on the table because having a rank specific to a single mod is pretty weird

**Comment by Rekunan** [comment]
@Okorin it's not only for a single mod, it's all mods that only impact visuals (ex: flashlight and fade in).

**Comment by Okorin** [comment]
Ttue but how does this invalidate my point

**Comment by Morroni** [comment]
> If this is the direction you're going with, I think it should be taken further by adding an FC count display to player profiles. I don't want the S rank counts on profiles to stray even further from the FC count than they currently are.

S rank doesn't show FC even now in stable osu!std. It just represents the play with decent accuracy and zero misscount. 

For instance, if you break combo on sliders, or even if you just loose some combo at slider ends, it will not be a FC anymore, and you will not receive your "Perfect" sign on your score screen, though it will still be a S rank play.

That's why it make no sense, and discussions about protecting "S-rank-aka-FC-badges" neither.

**Comment by Rekunan** [comment]
@Okorin cause it makes it perfectly justifiable and not weird for the distinction to exist in the first place.

**Comment by WitherFlower** [comment]
> S rank doesn't show FC even now in stable osu!std. It just represents the play with decent accuracy and zero misscount.

@Morroni I know, but I feel like that's what S ranks should have meant all along. I don't think any new player finds it logical that you can get an S rank while breaking combo. At least that's the situation I was in when I started. 

> That's why it doesn't make any sense, and discussions about protecting "S-rank-aka-FC-badges" neither.

I think FCs should be rewarded by something and I want to know how many FCs a player has without doing guesswork when going to their profile. I'm having trouble following what "doesn't make sense" here.

**Comment by ominoussage** [comment]
> > S ≥95%FC
> 
> Regarding stable, I'm pretty sure many people is not going to like this, but I consider it a necessary change if the 50 acc/50 combo or 70 acc/50 combo score change is happening. Of course it will be harder to get an S but it's the ideal chance to let the community know that accuracy will be more important towards the new 'meta' that lazer will bring.

What I don't like about it is why does it have to be an FC also? Take this for example, you got your S but it's only 96% but then there's a play in the leaderboard with a 99% A grade that only missed near the end and had more pp than your play.

The main problem here is that you got an S while he only got an A when the only wrong thing he did was missing. He had more pp and acc, but you still end up getting the better grade. Looking at it from a sort of statistics perspective, it rewards unfairly.

**Comment by bdach** [comment]
I'm hesitant to bring this thread up again, since there is a significant plurality of opinion, but given that [this comment](https://github.com/ppy/osu/issues/4184#issuecomment-1727923533) got some positive reactions, maybe we just take that proposal and run with it (and thus move this out of discussion)?

**Comment by peppy** [comment]
I think it's fine to go with this, yes.

**Comment by peppy** [comment]
> It seems a conclusion was reached above, but I'm not sure what, if anything, is actionable here. Care to explain what needs to be done in order to resolve this?

If we're going with my final proposal, at face value no changes need to be made... but I think we forgot about osu!catch, where having the cutoffs as low as they are right now makes grading too easy. So we likely want to adjust for osu!catch as per tables in https://github.com/ppy/osu/issues/4184#issuecomment-1724954327?

**Comment by smoogipoo** [comment]
Well, at least having simple cutoffs is better than the complex one the osu ruleset had in stable.
